### ISSUE TITLE
[CI] MlDistributedFailureIT testFullClusterRestart failing

### ISSUE THREAD

**Issue by bpintea** [elastic/elasticsearch#95947]
Same test as in https://github.com/elastic/elasticsearch/issues/87593 , but seems a different failure.

**Build scan:**
https://gradle-enterprise.elastic.co/s/yki7rvntlumim/tests/:x-pack:plugin:ml:internalClusterTest/org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT/testFullClusterRestart

**Reproduction line:**
```
./gradlew ':x-pack:plugin:ml:internalClusterTest' --tests "org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT.testFullClusterRestart" -Dtests.seed=C3442988E246C62B -Dtests.locale=sr-BA -Dtests.timezone=CNT -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT&tests.test=testFullClusterRestart

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=1214, name=elasticsearch[node_t1][system_write][T#1], state=RUNNABLE, group=TGRP-MlDistributedFailureIT]

  at __randomizedtesting.SeedInfo.seed([C3442988E246C62B:926B3899176A8E84]:0)

  Caused by: java.lang.AssertionError: [.ml-config/S-M0EHgdRMKOwkIr05e05g][[.ml-config][0]] org.elasticsearch.index.shard.IndexShardClosedException: CurrentState[CLOSED] operation only allowed when not closed

    at __randomizedtesting.SeedInfo.seed([C3442988E246C62B]:0)
    at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:310)
    at org.elasticsearch.index.engine.Engine.externalRefresh(Engine.java:1037)
    at org.elasticsearch.index.shard.IndexShard.externalRefresh(IndexShard.java:1239)
    at org.elasticsearch.action.support.replication.PostWriteRefresh.immediate(PostWriteRefresh.java:96)
    at org.elasticsearch.action.support.replication.PostWriteRefresh.refreshShard(PostWriteRefresh.java:62)
    at org.elasticsearch.action.support.replication.TransportWriteAction$AsyncAfterWriteAction.run(TransportWriteAction.java:505)
    at org.elasticsearch.action.support.replication.TransportWriteAction$WritePrimaryResult.runPostReplicationActions(TransportWriteAction.java:318)
    at org.elasticsearch.action.support.replication.ReplicationOperation.handlePrimaryResult(ReplicationOperation.java:173)
    at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:158)
    at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:93)
    at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:308)
    at org.elasticsearch.action.bulk.TransportShardBulkAction$2.finishRequest(TransportShardBulkAction.java:273)
    at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:235)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:286)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:137)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:74)
    at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:215)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
    at java.lang.Thread.run(Thread.java:1623)

    Caused by: org.elasticsearch.index.shard.IndexShardClosedException: CurrentState[CLOSED] operation only allowed when not closed

      at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:2187)
      at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:2181)
      at org.elasticsearch.index.shard.IndexShard.getReplicationGroup(IndexShard.java:2765)
      at org.elasticsearch.action.support.replication.PostWriteRefresh$2.onResponse(PostWriteRefresh.java:65)
      at org.elasticsearch.action.support.replication.PostWriteRefresh$2.onResponse(PostWriteRefresh.java:62)
      at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:308)
      at org.elasticsearch.index.engine.Engine.externalRefresh(Engine.java:1037)
      at org.elasticsearch.index.shard.IndexShard.externalRefresh(IndexShard.java:1239)
      at org.elasticsearch.action.support.replication.PostWriteRefresh.immediate(PostWriteRefresh.java:96)
      at org.elasticsearch.action.support.replication.PostWriteRefresh.refreshShard(PostWriteRefresh.java:62)
      at org.elasticsearch.action.support.replication.TransportWriteAction$AsyncAfterWriteAction.run(TransportWriteAction.java:505)
      at org.elasticsearch.action.support.replication.TransportWriteAction$WritePrimaryResult.runPostReplicationActions(TransportWriteAction.java:318)
      at org.elasticsearch.action.support.replication.ReplicationOperation.handlePrimaryResult(ReplicationOperation.java:173)
      at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:158)
      at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:93)
      at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:308)
      at org.elasticsearch.action.bulk.TransportShardBulkAction$2.finishRequest(TransportShardBulkAction.java:273)
      at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:235)
      at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
      at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:286)
      at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:137)
      at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:74)
      at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:215)
      at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
      at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
      at java.lang.Thread.run(Thread.java:1623)

```

**Comment by danielmitterdorfer** [comment]
Another instance of this happened in https://gradle-enterprise.elastic.co/s/fsxhpfod4bdmi/tests/:x-pack:plugin:ml:internalClusterTest/org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT/testFullClusterRestart?top-execution=1.

**Comment by HiDAl** [comment]
I got some failures in my PRs too 
https://gradle-enterprise.elastic.co/s/f6a5w5h532zku/tests/:x-pack:plugin:ml:internalClusterTest/org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT/testFullClusterRestart?top-execution=1

**Comment by droberts195** [comment]
I suspect this is a case of stateless changes spilling over into current product in a very subtle way.

The suite of tests where these failures are being seen is all about simulating network disruption.

When we write a document to the `.ml-config` system index with immediate refresh and the refresh part of the overall operation cannot happen because the network failure has cut off the required node, we now get an index closed exception. It seems like this didn't ever happen before May this year. So I suspect that a change was made to immediate refresh semantics in May that means inter-node communications of an index with immediate refresh now happens in two messages rather than one, so it's now possible for a network failure to occur in between the two messages.

**Comment by droberts195** [comment]
Failed again in exactly the same way in https://gradle-enterprise.elastic.co/s/ulxqjtna2trp2/tests/:x-pack:plugin:ml:internalClusterTest/org.elasticsearch.xpack.ml.integration.MlDistributedFailureIT/testFullClusterRestart?top-execution=1

I think it's a problem in core Elasticsearch that the error is manifesting as an uncaught exception:

```
juil. 03, 2023 7:35:05 PM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
WARNING: Uncaught exception in thread: Thread[elasticsearch[node_t1][system_write][T#2],5,TGRP-MlDistributedFailureIT]
java.lang.AssertionError: [.ml-config/RRS-QrHCRbmN1AnjJWZhAA][[.ml-config][0]] org.elasticsearch.index.shard.IndexShardClosedException: CurrentState[CLOSED] operation only allowed when not closed
	at __randomizedtesting.SeedInfo.seed([1E8C78115315E2CA]:0)
	at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:321)
	at org.elasticsearch.index.engine.Engine.externalRefresh(Engine.java:1051)
	at org.elasticsearch.index.shard.IndexShard.externalRefresh(IndexShard.java:1236)
	at org.elasticsearch.action.support.replication.PostWriteRefresh.immediate(PostWriteRefresh.java:101)
	at org.elasticsearch.action.support.replication.PostWriteRefresh.refreshShard(PostWriteRefresh.java:65)
	at org.elasticsearch.action.support.replication.TransportWriteAction$AsyncAfterWriteAction.run(TransportWriteAction.java:505)
	at org.elasticsearch.action.support.replication.TransportWriteAction$WritePrimaryResult.runPostReplicationActions(TransportWriteAction.java:318)
	at org.elasticsearch.action.support.replication.ReplicationOperation.handlePrimaryResult(ReplicationOperation.java:173)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:169)
	at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:94)
	at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:319)
	at org.elasticsearch.action.bulk.TransportShardBulkAction$2.finishRequest(TransportShardBulkAction.java:273)
	at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:235)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:286)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:137)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:74)
	at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:215)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: [.ml-config/RRS-QrHCRbmN1AnjJWZhAA][[.ml-config][0]] org.elasticsearch.index.shard.IndexShardClosedException: CurrentState[CLOSED] operation only allowed when not closed
	at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:2190)
	at org.elasticsearch.index.shard.IndexShard.verifyNotClosed(IndexShard.java:2184)
	at org.elasticsearch.index.shard.IndexShard.getReplicationGroup(IndexShard.java:2777)
	at org.elasticsearch.action.support.replication.PostWriteRefresh$2.onResponse(PostWriteRefresh.java:70)
	at org.elasticsearch.action.support.replication.PostWriteRefresh$2.onResponse(PostWriteRefresh.java:65)
	at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:319)
	... 24 more
```

The ML team could potentially workaround this problem if that exception were returned to the action listener.

Please could somebody from @elastic/es-distributed comment on the expected behaviour in this case? To summarise, the test is simulating what happens if there's network disruption, and it appears that the network disruption has struck in between the index and the refresh of an indexing request with immediate refresh. The index that's being reported as closed has not been explicitly closed, but is probably inaccessible momentarily during the test. The problem started happening in May, so a likely explanation is that some refactoring for stateless has changed the behaviour of current product.

**Comment by hendrikmuhs** [comment]
https://gradle-enterprise.elastic.co/s/ulxqjtna2trp2

**Comment by pxsalehi** [comment]
If I'm not mistaken, the fact that we check if the index is closed and fail if it is, is not new. But the failure should definitely be caught and propagated to the caller. I've created: https://github.com/elastic/elasticsearch/pull/97333, maybe @Tim-Brooks can have a look, and if that is what is missing then the ML caller should get the failure callback.

**Comment by droberts195** [comment]
> the fact that we check if the index is closed and fail if it is, is not new

The index is NOT closed. There is network disruption that prevents it being accessed.

> I've created: https://github.com/elastic/elasticsearch/pull/97333

The piece of code modified in that PR was also changed by https://github.com/elastic/elasticsearch/pull/94825, and the timing of that PR is about right for when this test started failing.

So I think there _has_ been a change to current product introduced by the stateless changes, albeit only in the obscure case of network disruption occurring in between the index and refresh parts of an immediate refresh bulk request.

What used to happen prior to #94825? Did the refresh part of an immediate refresh bulk request wait forever for the network to come back if there was network disruption in between the index part and the refresh part?

It's worth thinking this through because it won't just be internal code that's affected by this. We could get end user application programs also seeing index closed exceptions for the first time _on indices that aren't closed_ when there is transient network disruption in a cluster.

**Comment by pxsalehi** [comment]
> The index is NOT closed. There is network disruption that prevents it being accessed.

The log snippet above suggests that the index is closed. I understand that closing the index is not part of the test, and might be a side-effect of the network disruption, but the check there sees `IndexShardState.CLOSED`. As you mentioned, the PR that adds the timeouts might have surfaced this issue since before that we wouldn't timeout.

I think https://github.com/elastic/elasticsearch/pull/97333 is still valid, just because we should catch the exception and fail the listener, and I don't see another option there. If someone else has an idea why we shouldn't do that, please let me know.



Maybe @Tim-Brooks and @arteam could confirm or correct this, if I'm missing something.

**Comment by DaveCTurner** [comment]
> > The index is NOT closed. There is network disruption that prevents it being accessed.
> The log snippet above suggests that the index is closed.

You're using different senses of the word "closed". It is closed in the sense that the shard has been removed from this node's routing table, causing a call to `IndexShard#close`, but it is not closed in the sense that someone has called the `POST $INDEX/_close` API.

**Comment by DaveCTurner** [comment]
> We could get end user application programs also seeing index closed exceptions for the first time on indices that aren't closed when there is transient network disruption in a cluster.

I think this is not a new failure mode during network disruption, it's what I expect to happen when requests are directed to a node that no longer holds an active copy of a shard. `IndexShardClosedException` triggers a retry on the new primary, see `org.elasticsearch.action.support.replication.TransportReplicationAction#retryPrimaryException`.

**Comment by droberts195** [comment]
It's looking like on the main branch this has been fixed by #97333. It hasn't failed on the main branch since that PR was merged.

On the 8.9 branch this test is going to be muted by #97652.

So I'll close this issue for now. It can be reopened if the failure recurs.
